### DISCUSSION TITLE
Access Point and Internet Access simultaneously

### DISCUSSION THREAD

**Discussion by mikemiketymike** [micropython/micropython#12903] [General]
So I want my raspberry pi pico to act as both an access point and connect to another wifi network (and through that the internet) at the same time.

I can get two network interfaces if I use both network.WLAN(network.AP_IF) and network.WLAN(network.STA_IF), I can access the pico on the AP_IF interface and the pico can access addresses on the network.WLAN(network.STA_IF) network (via a socket), which is great. 
However, when I try to access anything on the internet this completely fails. I _think_ it's because the network code doesn't know how to route packets if there are multiple interfaces which is understandable because how would it know which interface to use?

My question is, is there a way to set default routing so that it'll route via the STA_IF network (if the address is not matched by the AP_IF network netmask)?

I found this in the code so I guess it's not supported?
```
            // If the active NIC is changing disable the active one first.
            // Note the host driver, firmware and ESP all support simultaneous AP/STA,
            // however modnetwork.c doesn't support routing traffic to different NICs
```

**Comment by sophiedegran** [comment]
In AP_IP you are on 192.168.4.1 so 192.168.4.x and no route

Normally in STA_IF you've got a route in your lan interface gateway declaration

**Reply by mikemiketymike** [reply]
> In AP_IP you are on 192.168.4.1 so 192.168.4.x and no route

Understood.

> Normally in STA_IF you've got a route in your lan interface gateway declaration

The routing in STA_IF only seems to work if it's the only interface.
I even tried binding a socket to the STA_IF ip before trying to connect to the internet and the connection still failed (although a connection to an ip on the STA_IF network succeeds).

**Reply by sophiedegran** [reply]
Under pi pico you can't have both AP and STA_IF active under ESP32 it work like a charm...

**Reply by mikemiketymike** [reply]
Well I got this to work in the end after looking in the source.

It looks like lwip doesn't do complex routing, it selects an interface by either matching the net mask of the interface to the destination ip, or, if that fails, it uses the default interface. There doesn't look to be a way to set the default interface via micropython but it's always set as the most recent interface added. So, if you add the AP_IP then add the STA_IP then the STA_IP will get selected when you're trying to get to an IP that's not on either of the directly connected networks, which is exactly what I want because we're never going to want to route to the internet via AP_IF.
[edit - grammar!]

**Reply by sophiedegran** [reply]
Yes you're right.. On rp2 as we have little ressource i prefer to close AP_IP enable STAT_IP send my data, email ..
and close to re-enable  AP_IP .. for me it's more safe..

**Reply by mikemiketymike** [reply]
Just an extra note in the unlikely case anyone is interested in this thread.
You can change this behaviour selecting the default netif in the cyw43-driver module.
This patch will change it such that the STA interface (if there is one) will always be the default netif (and so will route all traffic that doesn't match the AP net mask).

```
pi@raspberrypi:~/src/micropython/lib/cyw43-driver $ git diff
diff --git a/src/cyw43_lwip.c b/src/cyw43_lwip.c
index 79b2a90..8926de8 100644
--- a/src/cyw43_lwip.c
+++ b/src/cyw43_lwip.c
@@ -207,7 +207,11 @@ void cyw43_cb_tcpip_init(cyw43_t *self, int itf) {
     #error Unsupported
     #endif
     netif_set_hostname(n, CYW43_HOST_NAME);
-    netif_set_default(n);
+    if (itf == 0) {
+        // Set default netif to STA since it seems unlikely we'd ever want
+        // want it to be AP
+        netif_set_default(n);
+    }
     netif_set_up(n);
 
 //    #ifndef NDEBUG
```

**Comment by ssplatt** [comment]
> Under pi pico you can't have both AP and STA_IF active under ESP32 it work like a charm...

on on a pico w, if wifi provisioning is being attempted, what is the proper way to disable the AP before enabling the STA client mode? I've been trying different types of "cleanup" code, like `ap.disconnect() ; ap.active(False)` and even setting `ap = Null` after that. The `client.connect()` code (basically the same as all of the getting started examples) says it is doing the right thing (connecting to the proper real wifi, getting an IP, setting default gw and dns servers) but then any attempt to get something outside the local network returns `OSError: [Errno 113] EHOSTUNREACH` so it's like the routing table isn't getting properly updated, but I can't find anyway to output the routing table to debug that.

Is there an example of properly shutting down/removing/cleaning up the AP mode before continuing in STA/client mode on a Pico W?

**Reply by Beormund** [reply]
Having the same issue. I host a captive portal to enable end user to connect to ssid, show the STA ip address to end user then shut down captive portal/AP etc. But  then NTP requests fail with error 113 EHOSTUNREACH. My "fix" is to machine.reset() which is annoying.

**Comment by centralware** [comment]
For anyone wanting to have ACCESS POINT as well as STATION at the same time but not wanting to recompile MicroPython...  it CAN be done in raw PY and it's _**simple**_ to do...  it's just a redundant step to gather ifconfig() from WiFi and inject it into AP before having a working WiFi. (You can also recompile uPython as @mikemiketymike mentions above which works perfectly, but this is a way not to do so.)

NOTE: It only looks like a lot because of all of the example stuff to show both working.

```
import network
import socket
import urequests
from time import sleep

wifi_ssid = 'your_network_ssid'
wifi_password = 'your_network_password'

ap_ssid = 'MicroPython-AP'
ap_password = '123456789'

# --- TEMPORARY WIFI CONNETION --- #
wlan = network.WLAN(network.STA_IF)
wlan.active(True)
wlan.connect(wifi_ssid, wifi_password)
while not wlan.isconnected(): sleep(0.25)
GATEWAY = wlan.ifconfig()[2]    # <-- THIS IS THE STUFF
DNS = wlan.ifconfig()[3]        # <-- WE NEED TO FIX AP
wlan=None; del wlan             # <-- DELETE the connection when done!

# --- REAL ACCESS POINT --- #
ap = network.WLAN(network.AP_IF)
ap.config(essid=ap_ssid, password=ap_password)
IPADDR = ap.ifconfig()[0]
NETMASK = ap.ifconfig()[1]
new_config=(IPADDR, NETMASK, GATEWAY, DNS) # <-- INJECT WIFI's GATEWAY/DNS
ap.ifconfig(new_config)
ap.active(True)
print("A.P.:", ap.ifconfig())

# --- REAL WIFI CONNETION --- #
wlan = network.WLAN(network.STA_IF)
wlan.active(True)
wlan.connect(wifi_ssid, wifi_password)
while not wlan.isconnected(): sleep(0.25)
print("WiFi:", wlan.ifconfig())

# --- DOWNLOAD A TEXT FILE ONLINE USING OUR REAL WIFI --- #
url = 'https://example-files.online-convert.com/document/txt/example.txt'
response = urequests.get(url)
if response.status_code == 200:
    with open('download.txt', 'w') as f:
        f.write(response.text)
    print("File downloaded successfully")
else:
    print("Failed to download file")

# --- TAKE OUR TEXT FILE AND TURN IT INTO A WEB PAGE --- #
def web_page():
    with open("download.txt", "r") as file: text=file.read()
    text=text.replace("\n","<br>\n") # <-- Makes the text file more READABLE online
    html = f"<html><body>{text}</body></html>"
    return html

# --- CREATE WEB SERVER TO SERVE THAT TEXT FILE WITH OUR AP --- #
s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.bind(('', 80))
s.listen(5)
while True:
    conn, addr = s.accept()
    print('Got a connection from %s' % str(addr))
    request = conn.recv(1024)
    response = web_page()
    conn.send('HTTP/1.1 200 OK\n')
    conn.send('Content-Type: text/html\n')
    conn.send('Connection: close\n\n')
    conn.sendall(response)
    conn.close()
```
